app/code/Magento/Catalog/Model/Product/Compare/ListCompare.php

```diff
@@ -92,7 +103,10 @@ public function addProduct($product)
 
         if (!$item->getId()) {
             $item->addProductData($product);
-            $item->save();
+            $productId = $item->getProductId();
+            if ($productId && $this->_catalogFactory->create()->load($productId)->getId()) {
```

Comment:
Please use `\Magento\Catalog\Model\ResourceModel\Product::load` method instead of deprecated model's one.
Entities must not be responsible for their own loading.

Reply:
Is it safe to use `\Magento\Catalog\Api\ProductRepositoryInterface` then?

Sample:

Constructor
```
        array $data = [],
        \Magento\Catalog\Api\ProductRepositoryInterface $productRepository = null
$this->productRepository = $productRepository ?: \Magento\Framework\App\ObjectManager::getInstance()
            ->get(\Magento\Catalog\Api\ProductRepositoryInterface::class);

```

and call it like:
```
if ($productId && $this->productRepository->getById($productId)) {
                $item->save();
            }
```

Reply:
@adarshkhatri yep, repository is actually more preferable compared to resource model.

Reply:
used Product Repo to load product.

Reply:
IMO it's better to use resource model from the performance perspective, since we just need to make sure the entity exists :)
Nonetheless, it's not so critical.

Reply:
I ended up using search criteria builder.